packages/alfa-affine/src/transformation.ts

```diff
@@ -1,221 +1,291 @@
+    return new Transformation(multiply(this._matrix, transformation._matrix));
   }
 
-  export function rotate(a: number, u: Vector<3> = [0, 0, 1]): Transformation {
-    const [x, y, z] = Vector.normalize(u);
-
-    return [
-      [
-        cos(a) + x ** 2 * (1 - cos(a)),
-        x * y * (1 - cos(a)) - z * sin(a),
-        x * z * (1 - cos(a)) + y * sin(a),
-        0,
-      ],
-      [
-        y * x * (1 - cos(a)) + z * sin(a),
-        cos(a) + y ** 2 * (1 - cos(a)),
-        y * z * (1 - cos(a)) - x * sin(a),
-        0,
-      ],
-      [
-        z * x * (1 - cos(a)) - y * sin(a),
-        z * y * (1 - cos(a)) + x * sin(a),
-        cos(a) + z ** 2 * (1 - cos(a)),
-        0,
-      ],
-      [0, 0, 0, 1],
-    ];
+  public translate(tx: number, ty: number = 0, tz: number = 0): Transformation {
+    return this.apply(Transformation.translate(tx, ty, tz));
   }
 
-  export function skew(a: number, b: number): Transformation {
-    return [
-      [1, tan(a), 0, 0],
-      [tan(b), 1, 0, 0],
-      [0, 0, 1, 0],
-      [0, 0, 0, 1],
-    ];
+  public scale(sx: number, sy: number, sz: number = 1): Transformation {
+    return this.apply(Transformation.scale(sy, sx));
   }
 
-  export function project(d: number): Transformation {
-    return [
-      [1, 0, 0, 0],
-      [0, 1, 0, 0],
-      [0, 0, 1, 0],
-      [0, 0, -1 / d, 1],
-    ];
+  public skew(a: number, b: number): Transformation {
+    return this.apply(Transformation.skew(a, b));
   }
 
-  export interface Decomposed {
-    /**
-     * A vector representing [x, y, z] translation.
-     */
-    translate: Vector<3>;
-
-    /**
-     * A vector representing [x, y, z] scaling.
-     */
-    scale: Vector<3>;
-
-    /**
-     * A vector representing [xy, xz, yz] skewing.
-     */
-    skew: Vector<3>;
-
-    /**
-     * A quaternion representing [x, y, z, w] rotation.
-     */
-    rotate: Vector<4>;
+  public rotate(a: number, u?: [number, number, number]): Transformation {
+    return this.apply(Transformation.rotate(a, u));
+  }
 
-    /**
-     * A quaternion representing [x, y, z, w] perspective.
-     */
-    perspective: Vector<4>;
+  public project(d: number): Transformation {
+    return this.apply(Transformation.project(d));
   }
 
   /**
    * @see https://drafts.csswg.org/css-transforms-2/#decomposing-a-3d-matrix
    */
-  export function decompose(
-    transformation: Transformation
-  ): Option<Decomposed> {
-    if (transformation[3][3] === 0) {
+  public decompose(): Option<Transformation.Components> {
+    const m = this._matrix;
+
+    if (m[3][3] === 0) {
       return None;
     }
 
-    const t = Matrix.clone(transformation);
-
     for (let i = 0; i < 4; i++) {
       for (let j = 0; j < 4; j++) {
-        t[i][j] /= t[3][3];
+        m[i][j] /= m[3][3];
       }
     }
 
-    let perspective: Matrix<4, 4> | Vector<4> = Matrix.clone(t);
+    // Perspective
+    let p: Matrix | Vector = Matrix.clone(m);
```

Comment:
I do prefer the terseness of the short variable names _for this particular case_ as the decomposition code otherwise becomes very jagged. Mathematical notation is this strange exception where I personally value terseness and brevity more than anything else 🙈